theano/compile/builders.py

```diff
@@ -115,40 +64,88 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
+    def grad(self, inputs, output_grads):
+        if self.cached_grad_ops:
+            return self.grad_ops(inputs, output_grads)
+
+        grad_inps = self.internal_inputs + output_grads
+        upstream_grads = dict(izip(self.internal_outputs, output_grads))
+        if self.grad_ops is not None:
+            grad_ops_l = self.grad_ops
+            if isinstance(grad_ops_l, list):
+                assert len(grad_ops_l) <= len(self.internal_inputs)
+                if len(grad_ops_l) < len(self.internal_inputs):
+                    grad_ops_l += [None] * (
+                        len(self.internal_inputs) - len(grad_ops_l))
+                # It is normal if some inputs are not needed in order
+                # to compute the gradient, so we ignore them.
+                gs = [go if go else type(self)(
+                    grad_inps,
+                    (lambda g: g if g else (lambda *a:None))(
+                        theano.gradient.grad(
+                            cost=None,
+                            known_grads=upstream_grads,
+                            wrt=[inp],
+                            disconnected_inputs='ignore')
+                    ), on_unused_input='ignore'
+                ) for go, inp in izip(grad_ops_l, self.internal_inputs)]
+
+                # since OpFromGraphBase only accepts input sequence,
+                # additional filtering is needed
+                def grad_ops(inps, grds):
+                    # nonlocal gs, grad_ops_l
+                    return [(go(inps, grds) if ov else go(*(inps + grds)))
+                            for go, ov in izip(gs, grad_ops_l)]
+            else:
+                grad_ops = grad_ops_l
+            self.grad_ops = grad_ops
+        else:
```

Comment:
The `else` case could be handled by the code above. You already support only providing *some* gradient overrides, just add something like this at the top:
```python
if self.grad_ops is None:
    self.grad_ops = [None] * len(self.internal_inputs)
```
Then make sure the code above does the same your `else` case does (i.e., do a single `grad` call).